torch/csrc/deploy/deploy.cpp

```diff
@@ -0,0 +1,117 @@
+#include "torch/csrc/deploy/deploy.h"
+
+#ifndef _WIN32
+#include <dlfcn.h>
+#include <libgen.h>
+#include <unistd.h>
+#endif
+
+namespace torch {
+
+Package InterpreterManager::load_package(const std::string& uri) {
+  return Package(uri, this);
+}
+
+InterpreterSession MovableObject::acquire_session(
+    const Interpreter* on_this_interpreter) {
+  InterpreterSession I = on_this_interpreter
+      ? on_this_interpreter->acquire_session()
+      : pImpl_->manager_->acquire_one();
+  I.self = I.impl_->unpickle_or_get(pImpl_->object_id_, pImpl_->data_);
+  return I;
+}
+
+void MovableObjectImpl::unload(const Interpreter* on_this_interpreter) {
+  if (!on_this_interpreter) {
+    for (auto& interp : manager_->all_instances()) {
+      unload(&interp);
+    }
+    return;
+  }
+
+  InterpreterSession I = on_this_interpreter->acquire_session();
+  I.impl_->unload(object_id_);
+}
+
+MovableObjectImpl::~MovableObjectImpl() {
+  unload(nullptr);
+}
+
+void MovableObject::unload(const Interpreter* on_this_interpreter) {
+  pImpl_->unload(on_this_interpreter);
+}
+
+MovableObject InterpreterSession::create_movable(PythonObject obj) {
+  TORCH_CHECK(
+      manager_,
+      "Can only create a movable object when the session was created from an interpreter that is part of a InterpreterManager");
+  auto pickled = impl_->pickle(self, obj);
+  return MovableObject(std::make_shared<MovableObjectImpl>(
+      manager_->next_object_id_++, std::move(pickled), manager_));
+}
+
+#ifndef _WIN32
+static std::string find_interpreter() {
+  // lookup the path to libtorch_cpu.so using dladdr,
+  // passing any symbol from within the library gives the
+  // path to the shared library
+  Dl_info info;
+  dladdr((void*)find_interpreter, &info);
+  char buf[PATH_MAX];
+  TORCH_INTERNAL_ASSERT(!*stpncpy(buf, info.dli_fname, PATH_MAX));
```

Comment:
what an arcane expression- i think it means:  assert( not (pointing to (a null character) ) )  
where stpncpy returns a pointer to a null terminator if it succeeds?  what does it return a pointer to otherwise, a null pointer?

Reply:
That expression is my passive aggressive rant against C's string library where they can't manage to ensure size-limited string copy is null terminated unless you use this weird incantation. In the event the path is too long, the string is not null terminated ad stpncpy will return a valid pointer to a non-null character. Ill make the expression more clear when I land this.